### ISSUE TITLE
Docker 19.03 does not always honor -w [workdir]

### ISSUE THREAD

**Issue by hholst80** [moby/moby#41979]
Docker 19.03 has a weird behavior with -w (workdir) 

I would expect the following command to enter the container in the directory `/workspace`, but it enters in `/root` instead.

```
➜  tugboat git:(master) ✗ docker exec -it -w /workspace tugboat zsh
tugboat /root » echo $HOME
/workspace
tugboat /root » stat /root
  File: /root -> workspace
  Size: 9         	Blocks: 0          IO Block: 4096   symbolic link
Device: 41h/65d	Inode: 63014639    Links: 1
Access: (0777/lrwxrwxrwx)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2021-02-03 20:18:42.000000000 +0000
Modify: 2021-02-03 20:18:42.000000000 +0000
Change: 2021-02-03 20:18:42.881178213 +0000
 Birth: -
tugboat /root » 
^F%                                                                                                                                                                                                                  ➜  tugboat git:(master) ✗ docker version
Client:
 Version:           19.03.13
 API version:       1.40
 Go version:        go1.15.1
 Git commit:        4484c46
 Built:             Fri Oct  2 19:31:30 2020
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.8
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.12.17
  Git commit:       afacb8b7f0
  Built:            Wed Mar 11 01:30:32 2020
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          v1.2.13
  GitCommit:        7ad184331fa3e55e52b890ea95e65ba581ae3429
 runc:
  Version:          1.0.0-rc10
  GitCommit:        dc9208a3303feef5b3839f4323d9beb36df0a9dd
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
➜  tugboat git:(master) ✗ 
```

```
➜  tugboat git:(master) ✗ docker inspect tugboat       
[
    {
        "Id": "0e61f6a927be9407be52d714e0202028ca4edf2665ca61342431e7dddfa46070",
        "Created": "2021-02-03T20:19:59.759272955Z",
        "Path": "sleep",
        "Args": [
            "inf"
        ],
        "State": {
            "Status": "running",
            "Running": true,
            "Paused": false,
            "Restarting": false,
            "OOMKilled": false,
            "Dead": false,
            "Pid": 44299,
            "ExitCode": 0,
            "Error": "",
            "StartedAt": "2021-02-03T20:20:00.146608204Z",
            "FinishedAt": "0001-01-01T00:00:00Z"
        },
        "Image": "sha256:a9afd86e000baf24187535a0d76d7fa97807f8ff895aaca72fbdaa0b81cfc6dc",
        "ResolvConfPath": "/var/lib/docker/containers/0e61f6a927be9407be52d714e0202028ca4edf2665ca61342431e7dddfa46070/resolv.conf",
        "HostnamePath": "/var/lib/docker/containers/0e61f6a927be9407be52d714e0202028ca4edf2665ca61342431e7dddfa46070/hostname",
        "HostsPath": "/var/lib/docker/containers/0e61f6a927be9407be52d714e0202028ca4edf2665ca61342431e7dddfa46070/hosts",
        "LogPath": "/var/lib/docker/containers/0e61f6a927be9407be52d714e0202028ca4edf2665ca61342431e7dddfa46070/0e61f6a927be9407be52d714e0202028ca4edf2665ca61342431e7dddfa46070-json.log",
        "Name": "/tugboat",
        "RestartCount": 0,
        "Driver": "overlay2",
        "Platform": "linux",
        "MountLabel": "",
        "ProcessLabel": "",
        "AppArmorProfile": "",
        "ExecIDs": null,
        "HostConfig": {
            "Binds": [
                "workspace:/workspace:rw",
                "/var/run/docker.sock:/var/run/docker.sock:rw"
            ],
            "ContainerIDFile": "",
            "LogConfig": {
                "Type": "json-file",
                "Config": {}
            },
            "NetworkMode": "tugboat_default",
            "PortBindings": {},
            "RestartPolicy": {
                "Name": "unless-stopped",
                "MaximumRetryCount": 0
            },
            "AutoRemove": false,
            "VolumeDriver": "",
            "VolumesFrom": [],
            "CapAdd": [
                "SYS_ADMIN"
            ],
            "CapDrop": null,
            "Capabilities": null,
            "Dns": null,
            "DnsOptions": null,
            "DnsSearch": null,
            "ExtraHosts": null,
            "GroupAdd": null,
            "IpcMode": "private",
            "Cgroup": "",
            "Links": null,
            "OomScoreAdj": 0,
            "PidMode": "",
            "Privileged": false,
            "PublishAllPorts": false,
            "ReadonlyRootfs": false,
            "SecurityOpt": null,
            "UTSMode": "",
            "UsernsMode": "",
            "ShmSize": 67108864,
            "Runtime": "runc",
            "ConsoleSize": [
                0,
                0
            ],
            "Isolation": "",
            "CpuShares": 0,
            "Memory": 0,
            "NanoCpus": 0,
            "CgroupParent": "",
            "BlkioWeight": 0,
            "BlkioWeightDevice": null,
            "BlkioDeviceReadBps": null,
            "BlkioDeviceWriteBps": null,
            "BlkioDeviceReadIOps": null,
            "BlkioDeviceWriteIOps": null,
            "CpuPeriod": 0,
            "CpuQuota": 0,
            "CpuRealtimePeriod": 0,
            "CpuRealtimeRuntime": 0,
            "CpusetCpus": "",
            "CpusetMems": "",
            "Devices": [
                {
                    "PathOnHost": "/dev/fuse",
                    "PathInContainer": "/dev/fuse",
                    "CgroupPermissions": "rwm"
                }
            ],
            "DeviceCgroupRules": null,
            "DeviceRequests": null,
            "KernelMemory": 0,
            "KernelMemoryTCP": 0,
            "MemoryReservation": 0,
            "MemorySwap": 0,
            "MemorySwappiness": null,
            "OomKillDisable": false,
            "PidsLimit": null,
            "Ulimits": null,
            "CpuCount": 0,
            "CpuPercent": 0,
            "IOMaximumIOps": 0,
            "IOMaximumBandwidth": 0,
            "MaskedPaths": [
                "/proc/asound",
                "/proc/acpi",
                "/proc/kcore",
                "/proc/keys",
                "/proc/latency_stats",
                "/proc/timer_list",
                "/proc/timer_stats",
                "/proc/sched_debug",
                "/proc/scsi",
                "/sys/firmware"
            ],
            "ReadonlyPaths": [
                "/proc/bus",
                "/proc/fs",
                "/proc/irq",
                "/proc/sys",
                "/proc/sysrq-trigger"
            ],
            "Init": true
        },
        "GraphDriver": {
            "Data": {
                "LowerDir": "/var/lib/docker/overlay2/889ff6c704b1645e2772bc98c1f3e06a80c2d091425b1cad94b8abfe5b7bf318-init/diff:/var/lib/docker/overlay2/add2b14f86c1fff8e85c3c438ca5fc740daf6b8941aa51ffb0a963ab0940e167/diff:/var/lib/docker/overlay2/2a75279c165cb974b4e4bd7944c1e84009d1da1880aa56dc5d2fe9e0aa19941f/diff:/var/lib/docker/overlay2/6608578f6ad8f27ea94b7fb476489c9a61ba6a16882c9baecf3c64f79e15c4dd/diff:/var/lib/docker/overlay2/9c3ff69e86f9dbf2b2243f0dbe409dcdc1646e5ac38ae8ee03de6787a224e0d0/diff:/var/lib/docker/overlay2/2d4536afd994dc3833227eeba34c40b2abbb40ba1950fbd53c634c46d4fdbda8/diff:/var/lib/docker/overlay2/2ca6ab5256009997d0abd07615739861036464d7d3003cd04c3dcf18ff8f500f/diff:/var/lib/docker/overlay2/7ebb70ad89d8d81914a24a9055f41d757a532047f142ccdbfe9bd3e97232703b/diff:/var/lib/docker/overlay2/1db2c3f3d641a1a860d0f2726bfbcd933057427651fdd38b944970e6fcab4614/diff:/var/lib/docker/overlay2/81b2a2e182a6dd204ccfc8522d9995d23cec2536e204868935db04f7c529487d/diff:/var/lib/docker/overlay2/ec402abb8edf4807ad6328318333d91ce27f8ed89a6b24a0d82f00bef35c55a1/diff:/var/lib/docker/overlay2/e331be7082bc623046852a06a97d363e3eae6d2a8b427aceeba0bc1fbc1eceb9/diff:/var/lib/docker/overlay2/d2d5e3fc587cf4c5ac699a94743594587f703eda87937e967e05ed2377ed5e35/diff:/var/lib/docker/overlay2/53ca8506e8280c3964859e84306d6e368eaa6f2e403639362cd442207ac41e2f/diff",
                "MergedDir": "/var/lib/docker/overlay2/889ff6c704b1645e2772bc98c1f3e06a80c2d091425b1cad94b8abfe5b7bf318/merged",
                "UpperDir": "/var/lib/docker/overlay2/889ff6c704b1645e2772bc98c1f3e06a80c2d091425b1cad94b8abfe5b7bf318/diff",
                "WorkDir": "/var/lib/docker/overlay2/889ff6c704b1645e2772bc98c1f3e06a80c2d091425b1cad94b8abfe5b7bf318/work"
            },
            "Name": "overlay2"
        },
        "Mounts": [
            {
                "Type": "volume",
                "Name": "workspace",
                "Source": "/var/lib/docker/volumes/workspace/_data",
                "Destination": "/workspace",
                "Driver": "local",
                "Mode": "rw",
                "RW": true,
                "Propagation": ""
            },
            {
                "Type": "bind",
                "Source": "/var/run/docker.sock",
                "Destination": "/var/run/docker.sock",
                "Mode": "rw",
                "RW": true,
                "Propagation": "rprivate"
            }
        ],
        "Config": {
            "Hostname": "tugboat",
            "Domainname": "",
            "User": "",
            "AttachStdin": false,
            "AttachStdout": false,
            "AttachStderr": false,
            "Tty": false,
            "OpenStdin": false,
            "StdinOnce": false,
            "Env": [
                "TERM=xterm-256color",
                "HOME=/workspace",
                "LC_ALL=C.UTF-8",
                "DOCKER_BUILDKIT=1",
                "COMPOSE_DOCKER_CLI_BUILD=1",
                "PATH=/envs:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
            ],
            "Cmd": [
                "sleep",
                "inf"
            ],
            "Image": "omcr.io/tugboat:latest",
            "Volumes": {
                "/var/run/docker.sock": {},
                "/workspace": {}
            },
            "WorkingDir": "/workspace",
            "Entrypoint": null,
            "OnBuild": null,
            "Labels": {
                "com.docker.compose.config-hash": "16e01ee4cc903551987bb8885df79744c87b5c01897e8cd16948d2cf30738f25",
                "com.docker.compose.container-number": "1",
                "com.docker.compose.oneoff": "False",
                "com.docker.compose.project": "tugboat",
                "com.docker.compose.project.config_files": "docker-compose.yml",
                "com.docker.compose.project.working_dir": "/home/om/Documents/GitHub/tugboat",
                "com.docker.compose.service": "tugboat",
                "com.docker.compose.version": "1.27.4"
            }
        },
        "NetworkSettings": {
            "Bridge": "",
            "SandboxID": "f7cb80c8f1ba5f2d00e3e9469ced0a98c7593c09a71525b18152d83d2b78061d",
            "HairpinMode": false,
            "LinkLocalIPv6Address": "",
            "LinkLocalIPv6PrefixLen": 0,
            "Ports": {},
            "SandboxKey": "/var/run/docker/netns/f7cb80c8f1ba",
            "SecondaryIPAddresses": null,
            "SecondaryIPv6Addresses": null,
            "EndpointID": "",
            "Gateway": "",
            "GlobalIPv6Address": "",
            "GlobalIPv6PrefixLen": 0,
            "IPAddress": "",
            "IPPrefixLen": 0,
            "IPv6Gateway": "",
            "MacAddress": "",
            "Networks": {
                "tugboat_default": {
                    "IPAMConfig": null,
                    "Links": null,
                    "Aliases": [
                        "tugboat",
                        "0e61f6a927be"
                    ],
                    "NetworkID": "cad380d2e093b5ee523a874bf8c42d08d8c7cf507272db123900241e485b45b4",
                    "EndpointID": "b5debb9482513fa911206d5f51cd30975b9c1d174cc6b5aed7accf92672d0d21",
                    "Gateway": "172.17.0.1",
                    "IPAddress": "172.17.0.2",
                    "IPPrefixLen": 16,
                    "IPv6Gateway": "",
                    "GlobalIPv6Address": "",
                    "GlobalIPv6PrefixLen": 0,
                    "MacAddress": "02:42:ac:11:00:02",
                    "DriverOpts": null
                }
            }
        }
    }
]
```

**Comment by thaJeztah** [comment]
I think this is due to how symlinks are handled by shells; they can either "resolve" the symlink and use that as working-directory, or consider the path of the _symlink_ to be the working directory. (some related discussion https://github.com/fish-shell/fish-shell/issues/1957 (and https://github.com/fish-shell/fish-shell/issues/7206)

For testing, run an alpine container;

```bash
docker run -it --rm alpine
```

Inside the container, create a directory, and a symlink to that directory:

```bash
mkdir /target-dir
ln -s /target-dir/ /symlink-dir
```

Change to `/symlink-dir`, and print the working directory

```bash
cd /symlink-dir
pwd
# /symlink-dir
```

Now change to `/symlink-dir` again, but this time with the `-P` option, then print the working directory again;

```bash
cd -P /symlink-dir
pwd
# /target-dir
```

Also see the [cd(1p) man page](https://man7.org/linux/man-pages/man1/cd.1p.html)


```markdown
       -P        Handle the operand dot-dot physically; symbolic link
                 components shall be resolved before dot-dot components
                 are processed (see step 7. in the DESCRIPTION).
```

**Comment by thaJeztah** [comment]
From the above, I don't think this is a bug, so I'll go ahead and close, but I hope this helps, and feel free to continue the conversation

**Comment by hholst80** [comment]
I do not think what you described, although interesting, is applicable in this case. If anything, I was seeing the opposite behavior. (symlinks were NOT resolved as root -> workspace)

1) docker exec -w is set to /workspace
2) env HOME is set to /workspace
3) home directory in /etc/passwd is set to /root (and default shell is bash)

I tested it with bash and it does not have the same effect, there -w /workspace shows me ~ and pwd = /workspace as expected. So in any case its zsh that does something funny here not specific to docker. I tried to move out all my zsh dot files with no effect on the starting directory being /root so its fundamental zhell behavior. If I change the passwd homedir to / or anything else it will honor the -w option in docker exec.

Weird stuff. But thats on zsh not on docker. Thanks for looking into it. I didn't even think of looking at the shell as being part of the problem.
